component/azstorage/block_blob.go

```diff
@@ -653,6 +655,200 @@ func (bb *BlockBlob) WriteFromBuffer(name string, metadata map[string]string, da
+			return err
+		}
+		// case 2: given offset is within the size of the blob - and the blob consists of multiple blocks
+		// case 3: new blocks need to be added
+	} else {
+		modifiedBlockList, oldDataSize, exceedsFileBlocks := fileOffsets.FindBlocksToModify(offset, length)
+		// keeps track of how much new data will be appended to the end of the file (applicable only to case 3)
+		newBufferSize := int64(0)
+		// case 3?
+		if exceedsFileBlocks {
+			// get length of blockID in order to generate a consistent size block ID so storage does not throw
+			existingBlockId, _ := base64.StdEncoding.DecodeString(fileOffsets.BlockOffsetList[0].Id)
+			blockIdLength := len(existingBlockId)
+			appendOnly, newBufferSize = bb.createNewBlocks(modifiedBlockList, fileOffsets, offset, length, int64(blockIdLength))
+		}
+		if len(modifiedBlockList.BlockOffsetList) > 0 {
+			// buffer that holds that pre-existing data in those blocks we're interested in
+			oldDataBuffer := make([]byte, oldDataSize+newBufferSize)
+			if !appendOnly {
+				// fetch the blocks that will be impacted by the new changes so we can overwrite them
+				bb.ReadInBuffer(name, modifiedBlockList.BlockOffsetList[0].StartIndex, oldDataSize, oldDataBuffer)
+			}
+			// this gives us where the offset with respect to the buffer that holds our old data - so we can start writing the new data
+			blockOffset := offset - modifiedBlockList.BlockOffsetList[0].StartIndex
+			copy(oldDataBuffer[blockOffset:], data)
+			err := bb.stageAndCommitModifiedBlocks(name, blobURL, oldDataBuffer, modifiedBlockList, fileOffsets)
+			return err
+		}
+	}
+	return nil
+}
+
+// TODO: make a similar method facing stream that would enable us to write to cached blocks then stage and commit
+func (bb *BlockBlob) stageAndCommitModifiedBlocks(name string, blobURL azblob.BlockBlobURL, data []byte, modifiedBlockList, blockList *common.BlockOffsetList) error {
+	blockOffset := int64(0)
+	for _, blk := range modifiedBlockList.BlockOffsetList {
+		_, err := blobURL.StageBlock(context.Background(),
+			blk.Id,
+			bytes.NewReader(data[blockOffset:blk.Size+blockOffset]),
+			bb.blobAccCond.LeaseAccessConditions,
+			nil,
+			bb.downloadOptions.ClientProvidedKeyOptions)
+		if err != nil {
+			log.Err("BlockBlob::stageAndCommitModifiedBlocks : Failed to stage to blob %s at block %v (%s)", name, blockOffset, err.Error())
+			return err
+		}
+		blockOffset = blk.Size + blockOffset
+	}
+	var blockIDList []string
+	// TODO: we can probably clean up this for loop - move it to types method?
+	for _, blk := range blockList.BlockOffsetList {
+		blockIDList = append(blockIDList, blk.Id)
+	}
+	_, err := blobURL.CommitBlockList(context.Background(),
+		blockIDList,
+		azblob.BlobHTTPHeaders{ContentType: getContentType(name)},
+		nil,
+		bb.blobAccCond,
+		bb.Config.defaultTier,
+		nil, // datalake doesn't support tags here
+		bb.downloadOptions.ClientProvidedKeyOptions)
+	if err != nil {
```

Comment:
If this commit block list fails then entire write operation is failure and we loose the changes right?

Reply:
yea if the flush fails then the entire operation fails. Once again with our pipeline approach storage is not meant to maintain a state and therefore if something fails (just like any of the other operations on azstorage with two calls) then operation fails.